agate/agate/queue_reading/queue_reader.py

```diff
@@ -0,0 +1,83 @@
+from .tracking_models import Project, ProjectSite
+import json
+import logging
+from agate.queue_reading.ingestion_updater import IngestionUpdater
+
+from varys import Varys
+
+logger = logging.getLogger(__name__)
+
+
+class QueueReader:
+    """
+    Class to read the rabbitMQ queues, find any new messages.
+    It monitors 3 queue families, inbound-matched, inbound-to_validate-{project} and inbound-results-{project}-{site}
+
+    It learns about new projects and sites by monitoring the inbound-matched queue.
+
+    It delegates the actual interpretation of the message to the IngestionUpdater class
+    """
+
+    def update(self, varys_client: Varys) -> None:
+        """
+        This method reads the queues for new messages and interprets the information
+        """
+
+        # self._receive(varys_client, exchange="inbound-s3", update_lists=False, exchange_key="inbound-s3")
+
+        self._receive(varys_client, exchange="inbound-matched", update_lists=True, exchange_key="inbound-matched")
+
+        for project in Project.objects.all():
+            self._receive(
+                varys_client,
+                exchange=f"inbound-to_validate-{project.key}",
+                update_lists=False,
+                exchange_key="inbound-to-validate")
+
+        for project_site in ProjectSite.objects.all():
+            self._receive(
+                varys_client,
+                exchange=f"inbound-results-{project_site.key}",
+                update_lists=False,
+                exchange_key="inbound-results")
+
+    def _receive(self, varys_client: Varys, exchange: str, update_lists: bool, exchange_key: str):
+        messages = varys_client.receive_batch(exchange=exchange, queue_suffix="agate", timeout=1)
+
+        for m in messages:
+            try:
+                self._update_item_from_message(m, exchange_key)
+                if update_lists:
+                    self._update_lists(m)
+                varys_client.acknowledge_message(m)
+            except Exception:
+                varys_client.nack_message(m)
+                raise
```

Comment:
Just to check that I've interpreted the logic here correctly, we try to incorporate the message, including some failure modes (e.g. a `JSONDecodeError`) that would lead us to acknowledge the message without actually adding anything to the database. Then for any other failure that isn't already handled, we `raise`?

Also, this blanket `except Exception` doesn't apply to errors in `_update_lists` because that already has its own blanket `except Exception: pass` block.

Reply:
I'm honestly not sure, which I'd say means let's just go with whatever you devise now and we can refine it later.

I think the question is what the team wants to happen. Do we want to record such failures? Or discard them as irrelevant? I'm probably being overly cautious. Probably a logging message is fine, since an admin of some sort will still be able to sift through the logs if we sense something outside of the planned workflow is happening.

Reply:
I've gone with always acknowledging the messages (so it doesn't keep repeating on us); but letting the exception go uncaught, so django shows/logs it like any python exception at any other time.